csrc/transformer/softmax_kernels.cu

```diff
@@ -575,10 +671,15 @@ void launch_attn_softmax_backward_v2(T* out_grad,
     else if (seq_length <= 2048)
         softmax_backward_kernel_v2<T, 64>
             <<<grid_dim, block_dim, 0, stream>>>(out_grad, soft_inp, seq_length);
+    else if (seq_length <= 4096)
+        softmax_backward_kernel_v2<T, 128>
+            <<<grid_dim, block_dim, 0, stream>>>(out_grad, soft_inp, seq_length);
+    else if (seq_length <= 8192)
+        softmax_backward_kernel_v2<T, 256>
+            <<<grid_dim, block_dim, 0, stream>>>(out_grad, soft_inp, seq_length);
     else
-        throw std::runtime_error(
-            std::string("Special sequence length found in softmax backward, seq_length: ") +
-            std::to_string(seq_length));
+        softmax_backward_kernel_arbitrary_length<<<grid_dim, block_dim, 0, stream>>>(
+            out_grad, soft_inp, seq_length / (4 << ((sizeof(T) & 2) >> 1)));
```

Comment:
For `seq_length > 8192` and `seq_length % 8 != 0` (for `__half`) we'll still have slight correctness issues if I'm reading this correctly. For training this shouldn't be an issue typically, but should we throw a runtime error here just to be safe?

Reply:
That's a good point, will do that.

Reply:
also, the sequence will be padded here:
https://github.com/microsoft/DeepSpeed/blob/fix-softmax-backward/deepspeed/ops/transformer/transformer.py#L178
So, as you said there is no concern on the training side.

Reply:
If we're padding I don't see a need to add a runtime error.